crates/cli/cli/src/commands/clear/packets.rs

```diff
@@ -0,0 +1,128 @@
+        long = "chain",
+        required = true,
+        value_name = "CHAIN_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the chain to query"
+    )]
+    chain_id: ChainId,
+
+    #[clap(
+        long = "port",
+        required = true,
+        value_name = "PORT_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the port"
+    )]
+    port_id: PortId,
+
+    #[clap(
+        long = "channel",
+        alias = "chan",
+        required = true,
+        value_name = "CHANNEL_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the channel"
+    )]
+    channel_id: ChannelId,
+
+    #[clap(
+        long = "client",
+        required = true,
+        value_name = "CLIENT_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the client"
+    )]
+    client_id: ClientId,
+
+    #[clap(
+        long = "counterparty-chain",
+        required = true,
+        value_name = "COUNTERPARTY_CHAIN_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the counterparty chain to query"
+    )]
+    counterparty_chain_id: ChainId,
+
+    #[clap(
+        long = "counterparty-client",
+        required = true,
+        value_name = "COUNTERPARTY_CLIENT_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the counterparty client"
+    )]
+    counterparty_client_id: ClientId,
+
+    #[clap(
+        long = "counterparty-port",
+        required = true,
+        value_name = "COUNTERPARTY_PORT_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the counterparty port"
+    )]
+    counterparty_port_id: PortId,
+
+    #[clap(
+        long = "counterparty-channel",
+        required = true,
+        value_name = "COUNTERPARTY_CHANNEL_ID",
+        help_heading = "REQUIRED",
+        help = "Identifier of the counterparty channel"
+    )]
+    counterparty_channel_id: ChannelId,
+}
+
+impl CommandRunner<CosmosBuilder> for PacketsClear {
+    async fn run(&self, builder: &CosmosBuilder) -> Result<Output> {
+        let chain = builder.build_chain(&self.chain_id).await?;
+
+        let relayer = builder
+            .build_relay(
+                RelayAToBTarget,
+                &self.chain_id,
+                &self.counterparty_chain_id,
+                &self.client_id,
+                &self.counterparty_client_id, // nothing to pass here
+            )
+            .await?;
+
+        let target_height = chain.query_chain_height().await?;
+
+        relayer
+            .send_update_client_messages(DestinationTarget, &target_height)
+            .await
+            .wrap_err("Failed to send update client message")?;
+
+        <CosmosRelay as CanClearPackets>::clear_packets(
+            &relayer,
+            &self.channel_id,
+            &self.port_id,
+            &self.counterparty_channel_id,
+            &self.counterparty_port_id,
+        )
+        .await?;
+
+        Ok(Output::success("Packet clear"))
```

Comment:
Can we make `clear_packets` return the number of packets cleared? Or even a list of their sequence numbers? Then we can do:

```suggestion
        let packets_cleared = <CosmosRelay as CanClearPackets>::clear_packets(
            &relayer,
            &self.channel_id,
            &self.port_id,
            &self.counterparty_channel_id,
            &self.counterparty_port_id,
        )
        .await?;

        if json() {
          Ok(Output::success(json!({
            "total": packets_cleared.len(),
            "sequences": packets_cleared,
          })));
        } else {
          Ok(Output::success("Cleared {} packets: {}", packets_cleared.len(), PrettySlice(&packets_cleared)))
        }
```

Reply:
I think possibly a more flexible way to implement this is by making use of a better designed telemetry or logging abstraction to collect the result. Basically, the stats are simply collected as list of receive/ack relayed.

In the case of the packet clearing command, the result stats is what the same as the packet clearing execution. While in normal continuous execution, the stats will be continually updated together with the other relaying operations.

Reply:
> We would probably need to create a shared result state behind a mutex, and update the state when a clear packet task is completed.

Can't one instead collect the results yielded by all the concurrent tasks and merge them, ie. instead of using `for_each_concurrent`, use `flat_map_unordered` or something like that.

Whatever method we choose, let's keep it concrete for now and abstract over it only when necessary.

> We also need to decide whether such state is useful only in the CLI, or whether we need to also track this in the periodic packet clearer.

Yeah it would be nice have the clearing stats being output as well on periodic clears.